### PR TITLE
DCR tenant-wise configurations : Get and Patch endpoint implementation 

### PR DESCRIPTION
**Pull request by anjuchamantha**
## Purpose
This PR will have the `GET` and `PATHCH` implementations for the following DCR configurations in the server configurations api.

1. **enableFapiEnforcement** - If true, an application which is created through DCR endpoint will be a FAPI compliant application.
2. **authenticationRequired** - If true, authentication will be required for DCR create endpoint.
3. **ssaJwks** - This JWKS url will be used to validate the SSA coming in DCR create request.
4. **mandateSSA** - This will mandate software_statement param in DCR create request. If true, _ssaJwks_ should also be present.

Only the `REPLACE` operation will be supported and it will override the server-level configuration in deployment.toml. 

> **Note**: This PR is dependant on https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2344

Git Issue: https://github.com/wso2/product-is/issues/19274

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/resources/configs.yaml [hunk 1]
```diff
@@ -1177,6 +1251,46 @@ components:
...
+      type: object
+      properties:
+        clientAuthenticationRequired:
+          type: boolean
+          description: If false, the client authentication is not required for the DCR request, otherwise, the configured authentication mechanism will be used.
+          example: false
+        ssaJwks:
+          type: string
+          description: The JWKS endpoint to validate the SSA.
+          example: https://keystore.openbankingtest.org.uk/0015800001HQQrZAAX/oQ4KoaavpOuoE7rvQsZEOV.jwks
```
Comment (anjuchamantha) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/resources/configs.yaml:1264]: general url

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/DCRConnectorUtil.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+import org.wso2.carbon.identity.oauth.dcr.model.DCRConfiguration;
+
+import javax.ws.rs.core.Response;
+
+import static org.wso2.carbon.identity.api.server.configs.common.Constants.ErrorMessage.ERROR_CODE_DCR_CONFIG_SSA_MANDATE;
+import static org.wso2.carbon.identity.api.server.configs.common.Constants.ErrorMessage.ERROR_DCR_CONFIG_SERVICE_NOT_FOUND;
+
+
+/**
+ * Util class for DCR connector
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/DCRConnectorUtil.java:42]:
```suggestion
 * Util class for DCR connector.
```

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/DCRConnectorUtil.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+import org.wso2.carbon.identity.api.server.configs.v1.model.DCRConfig;
+import org.wso2.carbon.identity.oauth.dcr.exception.DCRMException;
+import org.wso2.carbon.identity.oauth.dcr.model.DCRConfiguration;
+
+import javax.ws.rs.core.Response;
+
+import static org.wso2.carbon.identity.api.server.configs.common.Constants.ErrorMessage.ERROR_CODE_DCR_CONFIG_SSA_MANDATE;
+import static org.wso2.carbon.identity.api.server.configs.common.Constants.ErrorMessage.ERROR_DCR_CONFIG_SERVICE_NOT_FOUND;
+
+
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/DCRConnectorUtil.java:40]: unnecessary line

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/DCRConnectorUtil.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                    .getDCRConfiguration();
+            return dcrConfigurationToDCRConfig(dcrConfiguration);
+        } else {
+            throw new DCRConfigException(ERROR_DCR_CONFIG_SERVICE_NOT_FOUND.message(),
+                    ERROR_DCR_CONFIG_SERVICE_NOT_FOUND.code());
+        }
+
+    }
+
+    public static void setDCRConfig(DCRConfig dcrConfig) throws
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/DCRConnectorUtil.java:67]: method java doc?
please check other public methods.

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/Constants.java [hunk 5]
```diff
@@ -125,7 +133,7 @@ public enum ErrorMessage {
                 "Unable to update remote logging  configs.",
                 "Server encountered an " +
                         "error while updating the remote logging configs."),
-        ERROR_CODE_ERROR_GETTING_REMOTE_LOGGING_CONFIGS("65008",
+        ERROR_CODE_ERROR_GETTING_REMOTE_LOGGING_CONFIGS("65017",
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/Constants.java:136]: Why did we change an already exposed error code?

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/Constants.java [hunk 6]
```diff
@@ -142,6 +150,26 @@ public enum ErrorMessage {
                 "Resource not found.",
                 "Unable to find a resource matching the provided log type %s."),
 
+        /**
+         * DCR Configuration errors.
+         */
+        ERROR_DCR_CONFIG_SERVICE_NOT_FOUND("60509",
```
Comment (DMHP) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/Constants.java:156]: We have provided error codes as **60509**, 65018, 65019, **650290** Are the bold colored error codes added mistakenly?

Comment (anjuchamantha) [reply components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/Constants.java]: Fixed with [18379bf](https://github.com/wso2/identity-api-server/pull/583/commits/18379bffe485d8f4d642a9dfc7ac2ffc6f1d8fdc)

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/gen/java/org/wso2/carbon/identity/api/server/configs/v1/ConfigsApi.java [hunk 7]
```diff
@@ -47,6 +49,7 @@
 import javax.ws.rs.*;
 import javax.ws.rs.core.Response;
 import io.swagger.annotations.*;
+import org.wso2.carbon.identity.api.server.configs.v1.model.Error;
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/gen/java/org/wso2/carbon/identity/api/server/configs/v1/ConfigsApi.java:52]: This is already defined in L33

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/gen/java/org/wso2/carbon/identity/api/server/configs/v1/model/DCRConfig.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+package org.wso2.carbon.identity.api.server.configs.v1.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+import javax.validation.constraints.*;
+
+
+import io.swagger.annotations.*;
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/gen/java/org/wso2/carbon/identity/api/server/configs/v1/model/DCRConfig.java:28]: Let's remove * imports and define the exact classes.

Comment (anjuchamantha) [reply components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/gen/java/org/wso2/carbon/identity/api/server/configs/v1/model/DCRConfig.java:28]: This class is auto generated from the [config.yaml](https://github.com/anjuchamantha/identity-api-server/blob/8ac3dde3cb0a93a18bc15cedbbc401a0841e25ae/components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/resources/configs.yaml)

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/gen/java/org/wso2/carbon/identity/api/server/configs/v1/model/DCRPatch.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+package org.wso2.carbon.identity.api.server.configs.v1.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+import javax.validation.constraints.*;
+
+
+import io.swagger.annotations.*;
```
Comment (RivinduM) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/gen/java/org/wso2/carbon/identity/api/server/configs/v1/model/DCRPatch.java:28]: Let's remove * imports and define the exact classes and order the imports alphabetically.

Comment (anjuchamantha) [reply components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/gen/java/org/wso2/carbon/identity/api/server/configs/v1/model/DCRPatch.java:28]: This class is auto generated from the [config.yaml](https://github.com/anjuchamantha/identity-api-server/blob/8ac3dde3cb0a93a18bc15cedbbc401a0841e25ae/components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/resources/configs.yaml)

If we do any changes, when someone else generate a new class from the schema, these classes may revert back to this state again.

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/factory/DCRMgtOGSiServiceFactory.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * Since this factory produces DCRConfigurationMgtService connector service,  there is a possibility that said
+ * connector not available in the distribution.
+ * So rather than designing as Factory Beans this class designed as Singleton.
+ */
+public class DCRMgtOGSiServiceFactory {
+
+    private static DCRConfigurationMgtService dcrConfigurationMgtService = null;
+    private static final Log log = LogFactory.getLog(DCRMgtOGSiServiceFactory.class);
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/factory/DCRMgtOGSiServiceFactory.java:34]:
```suggestion
    private static final Log LOG = LogFactory.getLog(DCRMgtOGSiServiceFactory.class);
```
please check other places too

Comment (anjuchamantha) [reply components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/factory/DCRMgtOGSiServiceFactory.java]: Fixed in https://github.com/wso2/identity-api-server/pull/583/commits/2f127c3263d4927dca8a8b683a23508f94a6b087

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/factory/DCRMgtOGSiServiceFactory.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    private static DCRConfigurationMgtService dcrConfigurationMgtService = null;
+    private static final Log LOG = LogFactory.getLog(DCRMgtOGSiServiceFactory.class);
+
+    /**
+     * This method return  the instance if the OSGi service exists.
+     * Else throw Null pointer Exception. We handle the exception gracefully.
+     *
+     * @return DCRConfigurationMgtService
+     */
+    public static DCRConfigurationMgtService getInstance() {
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/factory/DCRMgtOGSiServiceFactory.java:42]: do we really need this? cause DCR is not a connector right?

Comment (anjuchamantha) [reply components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.common/src/main/java/org/wso2/carbon/identity/api/server/configs/common/factory/DCRMgtOGSiServiceFactory.java]: Removed with [92937d8](https://github.com/wso2/identity-api-server/pull/583/commits/92937d817285e5deebedc0bae8c72c434bc91386)

File: components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/DCRConnectorUtil.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+package org.wso2.carbon.identity.api.server.configs.v1.function;
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/DCRConnectorUtil.java:20]: NEW LINE REQUIRED

Comment (anjuchamantha) [reply components/org.wso2.carbon.identity.api.server.configs/org.wso2.carbon.identity.api.server.configs.v1/src/main/java/org/wso2/carbon/identity/api/server/configs/v1/function/DCRConnectorUtil.java:19]: Fixed with [d700118](https://github.com/wso2/identity-api-server/pull/583/commits/d70011868554bcaed3d54d1c621510e1f53654cf)

**Review by Thumimku — APPROVED** [review]

**Pull request merged**